opal/mca/common/ucx/common_ucx.c

```diff
@@ -80,28 +82,38 @@ OPAL_DECLSPEC void opal_common_ucx_mca_var_register(const mca_base_component_t *
                                        &opal_common_ucx.opal_mem_hooks);
 
     if (NULL == opal_common_ucx.tls) {
-      opal_common_ucx.tls = default_tls;
+        // Extra level of string indirection needed to make ompi_info
+        // happy since it will unload this library before the MCA base
+        // cleans up the MCA vars. This will cause the string to go
+        // out of scope unless we place the pointer to it on the heap.
```

Comment:
Is this really the way that the MCA works?  I.e., isn't there some kind of safeguard to ensure that registered MCA vars are removed before their components are unloaded?

I ask because this seems like an odd restriction to place on components.  I.e., we have a system that allows a component to have an entirely segregated memory space (e.g., created/loaded by `dlopen()`), *except* for the MCA vars that it registers.  That's... weird.  Are we sure that that's the case?

Reply:
The problem here is with strings specifically, because `var_destruct()` (being called from opal_finalize_util) frees `var->mbv_storage->string` which refers to a pointer pointed from the common_ucx global space:
![image](https://github.com/open-mpi/ompi/assets/19361875/f26d2d53-acc1-4fd9-af8c-ebfd39fcc9b7)
 and no longer exists after the `ompi_info_close_components`.

Reply:
@bosilca @bwbarrett Is this how it's supposed to work?

Reply:
I think that's really a question for @hjelmn, not me.  This is what IBM came up with to make other components happy, and it definitely has been copied into the OFI common component.  But it seems like we've majorly borked something up if this is the intended usage model.

Reply:
Common components are weird IIRC. They don't necessarily have explicit register/open/close methods so they don't necessarily get cleaned up correctly. See:

https://github.com/open-mpi/ompi/blob/23113316aee7cb885ab078ac02998bfa800e82a5/opal/mca/base/mca_base_components_close.c#L31

Been awhile since I have worked on a common component but I believe you need to do that by hand. I know in some cases the registration method took an MCA component and tied the variables to that component. If you don't do that then you have to call that deregistration method on your own. Registration is automatic, deregistration is not in this case.

Reply:
Yeah, the issue is that this pattern may break MPI_T if the unload can happen when MPI_T is still usable. Any variable exposed with MPI_T would end up crashing. That is why we deregister all  the variables at once on unload.

Reply:
Ah.  What do you suggest, then?

Reply:
I suggest deregistering the group as is done in `mca_base_component_unload`

Reply:
@roiedanino Can you investigate @hjelmn's suggestion?

Reply:
Yes, I will

Reply:
@roiedanino Thanks!

I've moved the PR back to "Draft" so that we don't accidentally merge it yet.

Reply:
Tried it, didn't work